pkg/sources/registration.go

```diff
@@ -0,0 +1,195 @@
+package sources
+
+import (
+	"context"
+	"reflect"
+	"strings"
+
+	"github.com/sirupsen/logrus"
+	coreV1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/meta"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/tools/cache"
+
+	"github.com/kubeshop/botkube/pkg/config"
+	"github.com/kubeshop/botkube/pkg/utils"
+)
+
+type Registration struct {
+	informer        cache.SharedIndexInformer
+	log             logrus.FieldLogger
+	mapper          meta.RESTMapper
+	dynamicCli      dynamic.Interface
+	events          []config.EventType
+	mappedResources []string
+	mappedEvent     config.EventType
+}
+
+func (i Registration) handleEvent(ctx context.Context, resource string, target config.EventType, sourceRoutes []Route, fn eventHandler) {
+	switch target {
+	case config.CreateEvent:
+		i.informer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+			AddFunc: func(obj interface{}) {
+				sources := sourcesForObjNamespace(ctx, sourceRoutes, obj, i.log, i.mapper, i.dynamicCli)
+				i.log.Debugf("handle CreateEvent, resource: %s, sources: %+v", resource, sources)
+				if len(sources) > 0 {
+					fn(ctx, resource, sources, nil)(obj)
+				}
+			},
+		})
+	case config.DeleteEvent:
+		i.informer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+			DeleteFunc: func(obj interface{}) {
+				sources := sourcesForObjNamespace(ctx, sourceRoutes, obj, i.log, i.mapper, i.dynamicCli)
+				i.log.Debugf("handle DeleteEvent, resource: %s, sources: %+v", resource, sources)
+				if len(sources) > 0 {
+					fn(ctx, resource, sources, nil)(obj)
+				}
+			},
+		})
+	case config.UpdateEvent:
+		i.informer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+			UpdateFunc: func(oldObj, newObj interface{}) {
+				sources, diffs := qualifySourcesForUpdate(ctx, newObj, oldObj, sourceRoutes, i.log, i.mapper, i.dynamicCli)
+				i.log.Debugf("handle UpdateEvent, resource: %s, sources: %+v, diffs: %+v", resource, sources, diffs)
+				if len(sources) > 0 {
+					fn(ctx, resource, sources, diffs)(newObj)
+				}
+			},
+		})
+	}
+}
+
+func (i Registration) handleMapped(ctx context.Context, targetEvent config.EventType, routeTable map[string][]Entry, fn eventHandler) {
+	i.informer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			var eventObj coreV1.Event
+			err := utils.TransformIntoTypedObject(obj.(*unstructured.Unstructured), &eventObj)
+			if err != nil {
+				i.log.Errorf("Unable to transform object type: %v, into type: %v", reflect.TypeOf(obj), reflect.TypeOf(eventObj))
+			}
+			_, err = cache.MetaNamespaceKeyFunc(obj)
+			if err != nil {
+				i.log.Errorf("Failed to get MetaNamespaceKey from event resource")
```

Comment:
swallowed error, and redundant formatted error log

Reply:
I agree, while previously BotKube usually just logged errors, in new code we should return them as an output argument.

Reply:
True 👍 But I also saw in other places logging error instead of returning. Let's make sure we always return these errors, unless we can't 🙂

Reply:
For sure, we should always keep an eye for that. But, as @mszostok `AddFunc` stops this here.